### PR TITLE
Memos UI

### PR DESCRIPTION
**Pull request by banklesss**
Related to [YOMO-309](https://emurgo.atlassian.net/browse/YOMO-309)

**Create. Happy Path:**

https://user-images.githubusercontent.com/105349292/218822618-784d5f39-76b8-4000-aade-1d58a351aca4.mov

**Display. Happy Path:**

happy-path-display

**Error:**

https://user-images.githubusercontent.com/105349292/218109436-d49af0eb-6ac6-4e0e-b3b5-c91b964b0826.mov

**TX Detail without memo:**

image

[YOMO-309]: https://emurgo.atlassian.net/browse/YOMO-309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/TxHistory/TxDetails/TxDetails.tsx [hunk 1]
```diff
@@ -69,6 +69,12 @@ export const TxDetails = () => {
           </Boundary>
         </Banner>
 
+        <View style={styles.borderTop}>
+          <Label>{strings.memo}</Label>
+        </View>
+
+        <Text style={styles.assetsTitle}>{transaction.memo}</Text>
+
```
Comment (banklesss) [inline src/TxHistory/TxDetails/TxDetails.tsx]: I asked product about what happens when the memo is empty. Should the field be hidden?  shown empty field? default text ? etc

File: src/Send/MemoField.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        errorText={showError ? 'error' : undefined}
+        noErrors={true}
+      />
+
+      <View style={styles.helper}>
+        <Text style={[styles.warning, showError && styles.error]}>{strings.warning}</Text>
+
+        <Text style={[styles.counter, showError && styles.error]}>{`${memo.length}/${maxMemoValue}`}</Text>
+      </View>
+    </View>
```
Comment (banklesss) [inline src/Send/MemoField.tsx]: `HelperText` component from `react-native-paper` is limited so I created a custom helper.

File: src/TxHistory/TxDetails/TxDetails.tsx [hunk 3]
```diff
@@ -69,7 +69,15 @@ export const TxDetails = () => {
           </Boundary>
         </Banner>
 
-        <Label>{strings.fromAddresses}</Label>
+        <Label>{strings.memo}</Label>
+
+        <Text secondary monospace>
+          {transaction.memo !== null && transaction.memo.length > 0 ? transaction.memo : '-'}
```
Comment (banklesss) [inline src/TxHistory/TxDetails/TxDetails.tsx]: Product requirement.

File: src/components/TextInput/TextInput.tsx [hunk 4]
```diff
@@ -111,7 +111,19 @@ export const TextInput = React.forwardRef((props: Props, ref: ForwardedRef<RNTex
...
+            colors: {
+              background: COLORS.BACKGROUND,
+              placeholder: faded ? COLORS.GREY_6 : COLORS.TEXT_INPUT,
+              primary: faded ? COLORS.GREY_6 : COLORS.BLACK,
+              error: COLORS.ERROR_TEXT_COLOR,
+            },
+          }}
+          type={errorTextEnabled && !isEmptyString(errorText) ? 'error' : 'info'}
+          visible
+        >
```
Comment (banklesss) [inline src/components/TextInput/TextInput.tsx]: To unify styles

Before fixing it:
Screenshot 2023-02-07 at 14 41 22

File: src/Send/MemoField.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined}
+        noErrors={true}
+      />
+
+      <View style={styles.helper}>
+        <Text style={[styles.warning, showError && styles.error]}>{strings.warning}</Text>
+
+        <Text style={[styles.counter, showError && styles.error]}>{`${memo.length}/${maxMemoLength}`}</Text>
+      </View>
```
Comment (banklesss) [inline src/Send/MemoField.tsx]: https://github.com/Emurgo/yoroi-mobile/pull/2324#discussion_r1098017434

File: src/Send/MemoField.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  return (
+    <View style={styles.container}>
+      <TextInput
+        value={memo}
+        onChangeText={(memo) => memoChanged(memo.trim())}
+        label={strings.label}
+        autoComplete={false}
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined}
+        noErrors={true}
```
Comment (banklesss) [inline src/Send/MemoField.tsx]: Blocking `TextInput` helper but allowing the other input error features through `errorText`

File: src/Send/MemoField.tsx (hunk 5 above)
Comment (banklesss) [reply src/Send/MemoField.tsx]: Respecting the styles of the helpers

image

File: src/Send/MemoField.tsx (hunk 6 above)
Comment (wolverineks) [inline src/Send/MemoField.tsx]:
```suggestion
        noErrors
```

File: src/Send/MemoField.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        errorText={showError ? 'error' : undefined} // to show the error styling
+        noErrors={true} // to block the default helper
+        multiline={true}
+      />
+
+      <View style={styles.helper}>
+        <Text style={[styles.warning, showError && styles.error]}>{showError ? strings.error : strings.warning}</Text>
+
+        <Text style={[styles.counter, showError && styles.error]}>{`${memo.length}/${maxMemoLength}`}</Text>
+      </View>
```
Comment (banklesss) [inline src/Send/MemoField.tsx]: `HelperText` component from` react-native-paper` is limited so I created a custom helper.

File: src/Send/MemoField.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  return (
+    <View style={styles.container}>
+      <TextInput
+        value={memo}
+        onChangeText={(memo) => memoChanged(memo.trim())}
+        label={strings.label}
+        autoComplete={false}
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined} // to show the error styling
+        noErrors={true} // to block the default helper
```
Comment (wolverineks) [inline src/Send/MemoField.tsx]:
```suggestion
        noErrors // to block the default helper
```

File: src/Send/MemoField.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        multiline={true}
```
Comment (wolverineks) [inline src/Send/MemoField.tsx]:
```suggestion
        multiline
```

File: src/Send/MemoField.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+import {TextInput} from '../components'
+import {COLORS} from '../theme'
+import {useSend} from './Context/SendContext'
+
+export const maxMemoLength = 256
+
+export const MemoField = () => {
+  const strings = useStrings()
+  const {memo, memoChanged} = useSend()
```
Comment (stackchain) [inline src/Send/MemoField.tsx]: can it receive the `memo` and `memoChange` as props? `{memo: string, onChange: (text: string) =>}`

File: src/Send/MemoField.tsx [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined} // to show the error styling
+        noErrors={true} // to block the default helper
+        multiline={true}
+      />
+
+      <View style={styles.helper}>
+        <Text style={[styles.warning, showError && styles.error]}>{showError ? strings.error : strings.warning}</Text>
+
+        <Text style={[styles.counter, showError && styles.error]}>{`${memo.length}/${maxMemoLength}`}</Text>
```
Comment (stackchain) [inline src/Send/MemoField.tsx]: maybe abstract into `<Message />` and `<LengthCounter />` it's a little bloated

File: src/Send/MemoField.tsx [hunk 12]
```diff
@@ -0,0 +1,85 @@
+import React from 'react'
```
Comment (stackchain) [inline src/Send/MemoField.tsx]: for the file name may be just `Memo`, it would be nice if it represents more of the use case over the implementation

**Review by stackchain — COMMENTED** [review]
⬆️

File: src/Send/Memo.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    <View style={styles.container}>
+      <TextInput
+        value={memo}
+        onChangeText={(memo) => onChangeText(memo.trim())}
+        label={strings.label}
+        autoComplete={false}
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined} // to show the error styling
+        noHelper={true}
+        style={multilineMax && styles.input}
```
Comment (banklesss) [inline src/Send/Memo.tsx]: to meet the requirements

> it should be 1 row by default and then expand to max 3 rows

https://user-images.githubusercontent.com/105349292/218594223-6a989343-2e3f-4fcc-bf71-57448a360b20.mov

File: src/Send/Memo.tsx [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        multiline={true}
+      />
+
+      <View style={styles.helper}>
```
Comment (banklesss) [inline src/Send/Memo.tsx:33]: I've finally opted for an intermediate solution for a reason:

The TextInput helper has several dependencies on TextInput and has a specific style

Comment (banklesss) [reply src/Send/Memo.tsx:33]: Let me know if you think otherwise

File: src/Send/Memo.tsx [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+type Props = {
+  memo: string
+  onChangeText: (memo: string) => void
+}
+
+export const Memo = ({onChangeText, memo}: Props) => {
+  const strings = useStrings()
+
+  const showError = memo.length > maxMemoLength
+  const multilineMax = memo.length > 50
```
Comment (banklesss) [inline src/Send/Memo.tsx]: arbitrary number to meet requirements

https://github.com/Emurgo/yoroi-mobile/pull/2324#discussion_r1105108493

File: src/hooks/index.ts [hunk 16]
```diff
@@ -794,3 +794,18 @@ export const useResync = (wallet: YoroiWallet, options?: UseMutationOptions<void
...
   }
 }
+
+export const useSaveMemo = (
+  {wallet}: {wallet: YoroiWallet},
+  options?: UseMutationOptions<void, Error, {txId: string; memo: string}>,
+) => {
+  const mutation = useMutation({
+    mutationFn: ({txId, memo}) => wallet.saveMemo(txId, memo),
+    ...options,
```
Comment (wolverineks) [inline src/hooks/index.ts:803]: options should come before mutation fn

File: src/Send/Memo.tsx [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+const styles = StyleSheet.create({
+  container: {
+    paddingBottom: 25,
+  },
+  helper: {
+    flexDirection: 'row',
+    justifyContent: 'space-between',
+  },
+  input: {
+    height: 95,
```
Comment (stackchain) [inline src/Send/Memo.tsx]: shouldn't it be `maxHeight` ? otherwise the input will continue to expand

Comment (banklesss) [reply src/Send/Memo.tsx]: Here you can read the requirements 

https://github.com/Emurgo/yoroi-mobile/pull/2324#discussion_r1105108493

1 line default until 3 lines maximum. After that text scrolling

Comment (banklesss) [reply src/Send/Memo.tsx]: check the video please!

Comment (stackchain) [reply src/Send/Memo.tsx]: got it

**Review by stackchain — APPROVED** [review]

File: src/TxHistory/TxDetails/TxDetails.tsx (hunk 3 above)
Comment (wolverineks) [reply src/TxHistory/TxDetails/TxDetails.tsx]: `transaction.memo || '-'`

File: src/Send/SendScreen/SendScreen.tsx [hunk 18]
```diff
@@ -203,6 +216,8 @@ export const SendScreen = () => {
           />
         </TouchableOpacity>
 
+        <Memo memo={memo} onChangeText={memoChanged} />
```
Comment (wolverineks) [inline src/Send/SendScreen/SendScreen.tsx]: `MemoInput`

File: src/Send/Memo.tsx [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+      <TextInput
+        value={memo}
+        onChangeText={(memo) => onChangeText(memo.trim())}
+        label={strings.label}
+        autoComplete={false}
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined} // to show the error styling
+        noHelper={true}
+        style={multilineMax && styles.input}
+        multiline={true}
```
Comment (wolverineks) [inline src/Send/Memo.tsx]: multiline

File: src/Send/Memo.tsx [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+  return (
+    <View style={styles.container}>
+      <TextInput
+        value={memo}
+        onChangeText={(memo) => onChangeText(memo.trim())}
+        label={strings.label}
+        autoComplete={false}
+        testID="memoFieldInput"
+        errorText={showError ? 'error' : undefined} // to show the error styling
+        noHelper={true}
```
Comment (wolverineks) [inline src/Send/Memo.tsx]: noHelper

File: src/Send/Memo.tsx (hunk 13 above)
Comment (wolverineks) [inline src/Send/Memo.tsx]: what is this?

Comment (wolverineks) [reply src/Send/Memo.tsx]: `maxHeight` maybe?

File: src/Send/Memo.tsx (hunk 17 above)
Comment (wolverineks) [reply src/Send/Memo.tsx]: different screen sizes, different text sizes
could lead to unpredictable behavior

**Comment by banklesss** [comment]
> transaction.memo || '-'

eslint complains

**Comment by stackchain** [comment]
> > transaction.memo || '-'
> 
> eslint complains

`isEmptyString`

File: src/TxHistory/TxDetails/TxDetails.tsx [hunk 21]
```diff
@@ -33,6 +33,7 @@ export const TxDetails = () => {
   const [addressDetail, setAddressDetail] = React.useState<null | string>(null)
   const transactions = useTransactionInfos(wallet)
   const transaction = transactions[id]
+  const memo = !isEmptyString(transaction.memo) ? transaction.memo : '-'
```
Comment (wolverineks) [inline src/TxHistory/TxDetails/TxDetails.tsx:36]: ~transaction.memo ?? '-'~
transaction.memo || '-'

File: src/components/TextInput/TextInput.tsx [hunk 22]
```diff
@@ -48,13 +49,14 @@ export const TextInput = React.forwardRef((props: Props, ref: ForwardedRef<RNTex
   const {
     value,
     containerStyle,
+    renderComponentStyle,
```
Comment (wolverineks) [inline src/components/TextInput/TextInput.tsx:52]: textInputStyle? or just style?

**Review by wolverineks — APPROVED** [review]

File: src/TxHistory/TxDetails/TxDetails.tsx (hunk 21 above)
Comment (stackchain) [reply src/TxHistory/TxDetails/TxDetails.tsx:36]: lint :/

**Pull request merged**